File: sourcecode/ConvolveOpt.pas
program Convolve;

{$mode objfpc}{$H+}

uses SysUtils, CThreads, Classes, StopWatch;

type Timer = object
    private
        _beg, _end : LongInt; 
    public
        constructor Create;
        //destructor Destroy;
        function Elapsed : LongInt;
        procedure Reset;
end;

constructor Timer.Create;
begin
    _beg := GetTickCount64 mod High(LongInt);
end;

//destructor Timer.Destroy;
//begin
//    _end := GetTickCount64 mod High(LongInt);
//end;

function Timer.Elapsed : LongInt;
begin
    _end := GetTickCount64 mod High(LongInt);
    Result := _end - _beg;
end;

procedure Timer.Reset;
begin
    _beg := GetTickCount64 mod High(LongInt);
end;



Function f(n : Int64) : Real;
begin
    Result := 1.0 / n;
end;

Function g(n : Int64) : Real;
begin
    Result := sin(n);
end;

Function convolve(n : Int64) : Real;
Var 
    sum : Real;
    i   : Int64;
begin
    sum := 0;
    i := 1;
    while i*i <= n do
    begin
        if (n mod i = 0) then 
        begin
            sum := sum + f(i) * g(n div i);
            if (i*i <> n) then sum := sum + f(n div i) * g(i);
        end;
        i := i + 1;
    end;
    Result := sum;
end;

//Function doTest(n : Int64) : Int64;
//Var
//    a : Real;
//    t : Timer;
//begin
//    t.Reset;
//    a := convolve(n);
//    Result := t.Elapsed;
//end;

Function doTest(n : Integer) : Extended;
var
    a : Real;
    t : TStopWatch;
begin
    t := TStopwatch.Create;
    t.Start();
    a := convolve(n);
    t.Stop();
    Result := t.ElapsedTicks / 10000.0;
end;


begin
	write(doTest(StrToInt(ParamStr(1))):2:6);
end.

File: sourcecode/ConvolveOpt.cpp
#include <iostream>
#include <cmath>
#include <bits/stdc++.h>

using std::cout;
using std::setprecision;
using std::fixed;

class Timer
{
    public:
        Timer() { clock_gettime(CLOCK_REALTIME, &beg_); }
    
        double elapsed() {
            clock_gettime(CLOCK_REALTIME, &end_);
            return end_.tv_sec - beg_.tv_sec +
                (end_.tv_nsec - beg_.tv_nsec) / 1000000000.;
        }
    
        void reset() { clock_gettime(CLOCK_REALTIME, &beg_); }
    
    private:
        timespec beg_, end_;
};

double f(long int n) {
    return 1/n;
}

double g(long int n) {
    return sin(n);
}

double convolve(long int k) {
    double sum = 0.0;
    for (long int i = 1; i*i <= k; i++) {
        if (i < 0) break;
        if (k % i == 0) {
            sum += f(i) * g(k/i);
            if (i*i != k) sum += f(k/i) * g(i);
        }
    }
    return sum;
}

double doTest(int k) {
    double a;
    Timer t;
    t.reset();
    a = convolve(k);
    return t.elapsed()*1000.0;
}

int main(int argc, char** argv) {
	cout << setprecision(6) << fixed << doTest(atoi(argv[1]));
	return 0;
}
